spring/spring6/spring6-common/src/main/java/org/infinispan/spring/common/provider/SpringCache.java

```diff
@@ -47,6 +55,97 @@ public SpringCache(BasicCache nativeCache, long readTimeout, long writeTimeout)
       this.writeTimeout = writeTimeout;
    }
 
+
+   @Override
+   public CompletableFuture<?> retrieve(Object key) {
+      return encodedToValueWrapper(nativeCache.getAsync(key));
+   }
+   private CompletableFuture<ValueWrapper> encodedToValueWrapper(CompletableFuture<Object> cf) {
+      return cf.thenApply(value -> encodedToValueWrapper(value));
+   }
+
+   @Override
+   public <T> CompletableFuture<T> retrieve(Object key, Supplier<CompletableFuture<T>> valueLoaderAsync) {
+      CompletionStage<T> completionStage = handleAndCompose(nativeCache.getAsync(key), (v1, ex1) -> {
+         if (ex1 != null) {
+            return CompletableFuture.failedFuture(ex1);
+         }
+
+         if (v1 != null) {
+            return CompletableFuture.completedFuture(decodeNull(v1));
+         }
+
+         CompletableFuture result = new CompletableFuture<>();
+         CompletableFuture<T> computedValue = computationResults.putIfAbsent(key, result);
+         if (computedValue != null) {
+            return computedValue;
+         }
+
+         return handleAndCompose(valueLoaderAsync.get(), (newValue, ex2) -> {
+            if (ex2 != null) {
+               result.completeExceptionally(ex2);
+               computationResults.remove(key);
+            } else {
+               nativeCache.putIfAbsentAsync(key, encodeNull(newValue))
```

Comment:
I think we could avoid the operation if `newValue != null`?

Reply:
which operation ?

Reply:
Aaaah, I was just reading it wrong!! It is all good